### PR TITLE
Change how verbs provider reports devices from getinfo.

### PR DESCRIPTION
**Pull request by pmmccorm**
When returning fi_info structure for fi_getinfo, return -FI_ENODATA
when rdma_create_ep fails instead of returning its errno.

Seperate the endpoint creation from the fi_info structure allocation
so that the other places it is used do not have to free() unused
fi_info structures.

Iterate through all returned devices from rdma_getaddrinfo and create
corresponding linked fi_info structure. If any device returns error
 (except no memory) attempt ot cleanup and continue iterating.

Signed-off-by: pmmccorm <patrick.m.mccormick@intel.com>

### CODE REVIEW

**Comment by pmmccorm** [comment]
@jsquyres Can you test this branch and see if verbs still masks usnic ?

**Comment by jsquyres** [comment]
@goodell Is having a look at this.

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -612,58 +610,71 @@ fi_ibv_getepinfo(const char *node, const char *service,
...
+	ret = fi_ibv_getepinfo(node, service, flags, hints, &rai);
+
 	if (ret)
 		return ret;
 
-	rdma_destroy_ep(id);
-	return 0;
+	rhead = rai;
+
+	for (struct fi_info *tmp; rai; rai = rai->ai_next) {
```
Comment (jsquyres) [inline prov/verbs/src/fi_verbs.c]: Minor stylistic comment -- you can ignore if you don't like it: you might want to set rhead=rai as part of the loop.  E.g:

```c
for (struct fi_info *tmp, rhead = rai; rai; rai = rai->ai_next)
  ...
```

Comment (pmmccorm) [inline prov/verbs/src/fi_verbs.c]: Yes that does look better, will change when I rebase.

**Comment by goodell** [comment]
Code seems OK to me.  I can also confirm that that `fi_info` works correctly for usNIC devices with this change, but I think it will conflict with Sean's recent commits (e.g., bd99881) for issue #229 and #481.

**Comment by pmmccorm** [comment]
Yes Sean's is the better approach, this throws the error away.

**Pull request closed without merging**